### PR TITLE
Add exceptionsToTrace configuration support in @SentinelResource annotation

### LINKED ISSUE #540: Add configuration of exception classes in degrade rule
In some scenarios，this resource will be degraded when it throws a specific exception. 
So the configuration of exception class which should be traced (blocked), is needed.
For example:
  `
    private static final String RES_NAME = "RedisLockDegrade";

    @Resource
    private DsLock dsLock;

    @SentinelResource(value = RES_NAME, entryType = EntryType.IN, fallback = "fallback",
            exceptionClasses = {RedisLockException.class})
    public String execute(String arg) {
        return doBiz(arg);
    }

    /**
     * fallback
     * 降级函数，参数与原函数一致.
     */
    public String fallback(String arg) {
        dsLock.switchDbLock();
        try {
            System.out.println("Redis lock is degraded, use db lock instead");
            return doBiz(arg);
        } finally {
            dsLock.switchRedisLock();
        }
    }
`

### PR DESCRIPTION
**Pull request by beston123**
### Describe what this PR does / why we need it
In some scenarios，this resource will be degraded when it throws a specific exception.
So the configuration of exception class that should be traced (blocked), is needed.

### Does this pull request fix one issue?
Fixes #540 

### Describe how you did it
Add the configuration of exception classes that would be traced.

### Describe how to verify it
None

### Special notes for reviews
None

### CODE REVIEW

File: sentinel-core/src/main/java/com/alibaba/csp/sentinel/annotation/SentinelResource.java [hunk 1]
```diff
@@ -63,4 +59,9 @@
      * @return name of the fallback function, empty by default
      */
     String fallback() default "";
+
+    /**
+     * @return the exception classes to trace, Throwable.class by default
+     */
+    Class<? extends Throwable>[] exceptionClasses() default {Throwable.class};
```
Comment (sczyh30) [inline sentinel-core/src/main/java/com/alibaba/csp/sentinel/annotation/SentinelResource.java]: `exceptionClasses` might not be clear enough. Maybe it's better to be `exceptionsToTrace`?

Comment (beston123) [reply sentinel-core/src/main/java/com/alibaba/csp/sentinel/annotation/SentinelResource.java]: Agree with your proposal.

File: sentinel-extension/sentinel-annotation-aspectj/src/main/java/com/alibaba/csp/sentinel/annotation/aspectj/SentinelResourceAspect.java [hunk 2]
```diff
@@ -59,12 +59,33 @@ public Object invokeResourceWithSentinel(ProceedingJoinPoint pjp) throws Throwab
...
+     * @param ex
+     * @param exceptionsToTrace
+     * @return
+     */
+    private boolean isTrackedException(Throwable ex, Class<? extends Throwable>[] exceptionsToTrace) {
+        if (exceptionsToTrace == null) {
+            return false;
+        }
+        for (Class exceptionToTrace : exceptionsToTrace) {
+            if (ex.getClass().isAssignableFrom(exceptionToTrace)) {
```
Comment (sczyh30) [inline sentinel-extension/sentinel-annotation-aspectj/src/main/java/com/alibaba/csp/sentinel/annotation/aspectj/SentinelResourceAspect.java]: It's incorrect. For each exception to trace, it should be:

```suggestion
            if (exceptionToTrace.isAssignableFrom(ex.getClass())) {
```

**Please be sure to add test cases and run the demo to verify it.**

File: sentinel-extension/sentinel-annotation-aspectj/src/main/java/com/alibaba/csp/sentinel/annotation/aspectj/SentinelResourceAspect.java [hunk 3]
```diff
@@ -59,12 +59,33 @@ public Object invokeResourceWithSentinel(ProceedingJoinPoint pjp) throws Throwab
         } catch (BlockException ex) {
             return handleBlockException(pjp, annotation, ex);
         } catch (Throwable ex) {
-            Tracer.trace(ex);
+            if (isTrackedException(ex, annotation.exceptionsToTrace())) {
```
Comment (sczyh30) [inline sentinel-extension/sentinel-annotation-aspectj/src/main/java/com/alibaba/csp/sentinel/annotation/aspectj/SentinelResourceAspect.java]: Maybe we can move the logic to `AbstractSentinelAspectSupport` class to support extension for sub-classes. E.g.:

```java
protected void traceException(Throwable ex, SentinelResource annotation)
```

**Comment by sczyh30** [comment]
Thanks for contributing. I'll help to update the test cases and refactor code later. Please note that every new feature or enhancement **must** carry corresponding test cases.

**Comment by sczyh30** [comment]
Improved in https://github.com/alibaba/Sentinel/commit/61c525076fb057c73a18f3b3ea00ef738fc9c839

**Pull request merged**